core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -0,0 +1,200 @@
+  using Analysis =
+      FunctorAnalysis<FunctorPatternInterface::REDUCE, Policy, FunctorType>;
+  using execution_space = typename Analysis::execution_space;
+  using value_type      = typename Analysis::value_type;
+  using pointer_type    = typename Analysis::pointer_type;
+  using reference_type  = typename Analysis::reference_type;
+
+  using WorkTag = typename Policy::work_tag;
+  using ReducerConditional =
+      Kokkos::Impl::if_c<std::is_same<InvalidType, ReducerType>::value,
+                         FunctorType, ReducerType>;
+  using ReducerTypeFwd = typename ReducerConditional::type;
+  using WorkTagFwd =
+      std::conditional_t<std::is_same<InvalidType, ReducerType>::value, WorkTag,
+                         void>;
+  using ValueInit =
+      typename Kokkos::Impl::FunctorValueInit<ReducerTypeFwd, WorkTagFwd>;
+
+ public:
+  // V - View
+  template <typename V>
+  ParallelReduce(
+      const FunctorType& f, const Policy& p, const V& v,
+      typename std::enable_if<Kokkos::is_view<V>::value, void*>::type = nullptr)
+      : m_functor(f), m_policy(p), m_result_ptr(v.data()) {
+    // FIXME_SYCL custom reducer not yet implemented
+    if (m_result_ptr == nullptr)
+      Kokkos::abort("Custom reducer not yet implemented for SYCL backend");
+  }
+
+  ParallelReduce(const FunctorType& f, const Policy& p,
+                 const ReducerType& reducer)
+      : m_functor(f),
+        m_policy(p),
+        m_reducer(reducer),
+        m_result_ptr(reducer.view().data()) {
+    // FIXME_SYCL custom reducer not yet implemented
+    if (m_result_ptr == nullptr)
+      Kokkos::abort("Custom reducer not yet implemented for SYCL backend");
+  }
+
+ private:
+  template <typename TagType>
+  std::enable_if_t<std::is_void<TagType>::value> exec(reference_type update) {
+    using member_type = typename Policy::member_type;
+    member_type e     = m_policy.end();
+    for (member_type i = m_policy.begin(); i < e; ++i) m_functor(i, update);
+  }
+
+  template <typename TagType>
+  std::enable_if_t<!std::is_void<TagType>::value> exec(reference_type update) {
+    using member_type = typename Policy::member_type;
+    member_type e     = m_policy.end();
+    for (member_type i = m_policy.begin(); i < e; ++i)
+      m_functor(TagType{}, i, update);
+  }
+
+  template <typename PolicyType, typename Functor>
+  void sycl_direct_launch(const PolicyType& policy,
+                          const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    cl::sycl::queue& q = *instance.m_queue;
+
+    q.submit([this, functor, policy](cl::sycl::handler& cgh) {
+      // FIXME_SYCL a local size larger than 1 doesn't work for all cases
+      cl::sycl::nd_range<1> range(policy.end() - policy.begin(), 1);
+
+      value_type identity = 0;
+      ValueInit::init(functor, m_result_ptr);
```

Comment:
According to SYCL 2020 Provisional, if `m_result_ptr` is in `sycl::usm::alloc::device` memory, we cannot initialize `m_result_ptr` in this way, because the memory is not accessible on the host.  However, I don't think the current USM implementation enforces this restriction.

In other words, it is fine for right now, but I do expect this to break.  The better way to do this is to duplicate the logic that `ValueInit::init` does while using `q.memcpy(...)` to set the actual value.

Also, should we add a check ensuring that `m_result_ptr` is in either device or shared memory?  Without it, the resultant error is unhelpful in diagnosing what went wrong (an exception is thrown when `cgh.parallel_for` is called).